src/pmi/examples/test_pmi1.c

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+/* Simple example using PMI-v1 interface.
+ *
+ * To compile:
+ *     cc test_pmi1.c -lpmi
+ *
+ * To run:
+ *     mpirun -n 2 ./a.out
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "pmi.h"
+
+int main(int argc, char **argv)
+{
+
+    char kvsname[1024];
+    int has_parent, rank, size;
+
+    PMI_Init(&has_parent);
+    PMI_Get_rank(&rank);
+    PMI_Get_size(&size);
+    fprintf(stdout, "    :has_parent=%d, rank=%d, size=%d\n", has_parent, rank, size);
+
+    PMI_KVS_Get_my_name(kvsname, 1024);
+
+    if (rank == 0) {
+        /* NOTE: PMI-v1 does not allow space in both key and value */
+        PMI_KVS_Put(kvsname, "key0", "got-key0");
```

Comment:
It's a no-op in Simple PMI, but I think technically we should have a `PMI_KVS_Commit()` here.

Reply:
Yeah, I meant to discuss this --

PMI_KVS_Commit() flushed the local put messages to global KVS space, so it is useful if we want some sort of p2p kvs exchange where one process `Put-Commit` and another process `Get (wait)`. It is useful since it avoids global `PMI_Barrier`. On the other hand, if the application is to do global KVS exchange, a single PMI_Barrier is more efficient than individual processes each try to push their own message (on to other processes' KVS). That is, adding `PMI_KVS_Commit` before a `PMI_Barrier` can be costly if we do not implement it as a "no-op" (one of my plan).

Is there text somewhere guarantee a `PMI_KVS_Get` will get the message of a `PMI_KVS_Put` before a barrier? If there is, then it is legally okay to omit `PMI_KVS_Commit` here.

Reply:
The Notes section for `PMI_KVS_Put` in pmi.h says

> This function puts the key/value pair in the specified keyval space.  The
value is not visible to other processes until 'PMI_KVS_Commit()' is called.

So I think we are stuck requiring it before any other process can get the value. I do think as things are defined, a barrier could be avoided if the getting process just keeps calling `PMI_KVS_Get` until the value is available.

Reply:
We can treat those text as "standard" and see it as "stuck", or, since there was never an official standard, we can modify the text now :) . As long as the changes are backward compatible, I would think it'll be okay. What do you think?

Reply:
The word I want to add is a note to the `PMI_Barrier`, that `PMI_Barrier` automatically applies `PMI_KVS_Commit` for each process.

Reply:
I see. I think that is essentially the idea behind the semantic of `PMI2_KVS_Fence`. Then again, PMI2 doesn't have a non-collective API for making KV pairs visible.

The quoted text exists in various implementations, though I'd bet `PMI_KVS_Commit` is a no-op in all those cases (Slurm ✅  Microsoft ✅  Flux ✅  Cray ❓ ). Flux actually takes it a step further and states that KV pairs are not visible until after _both_ a commit and a barrier https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_13.html#key-value-store.

The PMI server in Hydra does not make any keys available until after a barrier, so I guess it's an OK change to make 🤷 .